### DISCUSSION TITLE
Lifetime of a signal

### DISCUSSION THREAD

**Discussion by elliotwaite** [solidjs/solid#1776] [Q&A] [answered]
Can signals always be read from or written to as long as they can be referenced? Or do they become unusable at some point before they get garbage collected?

For example, in this component, a `mounted` variable used to keep track of if the component is still mounted, and `setData()` is only called if the component is still mounted:
```
function MyComponent() {
  const [data, setData] = createSignal("")

  onMount(() => {
    let mounted = true
    
    const fn = async () => {
      const newData = await getData()
      if (mounted) {
        setData(newData)
      }
    }
    void fn()
    
    onCleanup(() => {
      mounted = false
    })
  })
  
  return <div>{data()}</div>
}
```
...but this seems to work just as well without errors:
```
function MyComponent() {
  const [data, setData] = createSignal("")

  onMount(() => {
    const fn = async () => {
      const newData = await getData()
      setData(newData)
    }
    void fn()
  })

  return <div>{data()}</div>
}
```

But is this safe? Can I always use `setData()` even if the component is no longer mounted?

**Comment by snnsnn** [comment, marked as answer]
Signal's logic is scoped to `createSignal` and `createSignal` returns a pair of regular JavaScript functions for getting and setting values. Since they are regular JavaScript functions regular JavaScript rules apply. There are no magic, no compiler tricks, no additional rules. 

That being said your logic is a bit weird. You can make API calls directly inside the component's body. onMount is there if you need to imperatively interact with the underlying DOM elements using refs. Refs will be bound to elements after components are rendered, meaning a ref will be undefined inside the components body, but will be available when onMount effect is executed.  It is an effect that executes once elements became available, does not mount anything. Other than that you can write everything directly into component's body.

You can check this answer for a detailed explanation on making API calls in solid: https://stackoverflow.com/a/74590574/7134134

**Reply by elliotwaite** [reply]
Got it. Thanks for the explanation and suggestions. I think I was falling back into React habits by using that unnecessary `onMount`.
